Fix pipeline state logic and worker thread error handling

colinsurprenant opened the pull request:
Relates to #12005 and fixes the issues in (2).
Should fix #10611

This PR fixes the following issues:

- A pipeline in the process of being created was not marked as such in the pipeline registry resulting in a situation where a slow to initialize pipeline would be recreated on state convergence resulting in a PQ `LockException` because that pipeline was already existing and held the PQ lock. 
- The worker threads were not correctly monitored for a worker loop exception resulting in a complete logstash crash upon any exception even when multiple pipelines are running. Now only the failed pipeline is terminated. If pipeline reloading is enabled, it is possible to edit the config and have that failed pipeline reloaded.

### TODO
- [x] review  build errors
- [X] make sure same logic is applied to the Ruby pipeline - followup issue created in #12106

yaauie commented on logstash-core/lib/logstash/pipelines_registry.rb:
```diff
@@ -17,40 +17,99 @@
+
+    def each_with_object(init, &block)
+      states = @lock.synchronize { @states.dup }
+      states.each_with_object(init, &block)
```
This makes me a little nervous, because I cannot tell what the expectation is here from an exclusivity standpoint. We guarantee that the caller yields elements from a snapshot acquired under mutual exclusion, but do we need to encapsulate that in case the caller _mutates_ any of the yielded items?

colinsurprenant replied:
Not really.  Each of the state objects are thread safe. This is simply to grab a collection copy of all the state objects at this point in time. Once that collection copy is grabbed, it is safe to iterate and read the state objects which are thread safe themselves. There is no need to lock the states collection while we iterate on it.

yaauie commented on logstash-core/lib/logstash/pipelines_registry.rb:
```diff
@@ -17,40 +17,99 @@
+
+    def remove(pipeline_id)
+      @lock.synchronize do
+        @states.delete(pipeline_id)
```
do we also need to dereference the `@locks[pipeline_id]`, or is the unbounded growth of per-pipeline lock cache safe enough because how many distinct pipelines can a process run anyway? If we do clean up here, we also have to mark the existing lock object as invalid, which is whole new can of worms.

colinsurprenant replied:
Nevermind my previous comment - we should definitely also delete from `@locks` ... this was an oversight.  Fixed.

yaauie replied:
Now that `PipelineStates#remove(pipeline_id)` deletes locks, _unless we have some other mutual exclusion*_, it is possible that the lock we acquire with `PipelineStates#get_lock` (e.g., at the top of `PipelinesRegistry#create_pipeline`, `PipelinesRegistry#terminate_pipeline`, `PipelinesRegistry#reload_pipeline`, etc.) can be deleted by another thread without the lock-holder's knowledge, and a new lock can be created in its place.

*: I _believe_ that `PipelinesRegistry#create_pipeline` and friends ares only called when the agent's `@convergence_lock` is exclusively held (by way of the various `PipelineAction`s), but there are a number of layers of indirection and I haven't been able to pin it down to prove it. Even if they do though, we should find a way to make this more explicit lest new logic be introduced that accidentally circumvents the mutual exclusion we currently enjoy.

An alternative would be to never actually give a reference the lock out, instead implementing `PipelineStates#with_lock(pipeline_id)` that yields to a provided block after acquiring the lock.

yaauie commented on logstash-core/lib/logstash/pipelines_registry.rb:
```diff
@@ -17,40 +17,99 @@
+    end
+
+    def get_lock(pipeline_id)
+      @lock.synchronize do
+        @locks[pipeline_id] ||= Mutex.new
+      end
```
Optional: avoid acquiring a lock when the value is already populated

```suggestion
      @locks.fetch(pipeline_id) do |h,k|
        @lock.synchronize do
          h[k] ||= Mutex.new
        end
      end
```

***

Nitpick: we seem to have two concerns that each require their own mutual exclusion, and could benefit from having separate locks for state-operations and lock-acquisition.

colinsurprenant replied:
In general the very first thing you have to do is to acquire the global lock because when `synchronize` is not yet called by this thread, the *happens-before* relationship is not established and what this thread sees in the `@locks` collection might not reflect what other thread have written to it.

Your suggestion should actually be 

```ruby
      @locks.fetch(pipeline_id) do |k|
        @lock.synchronize do
          @locks[k] ||= Mutex.new
        end
      end
```

But this will only work because you actually do the `||= Mutex.new` within the `synchronize` block. As I mentioned above, the actual state of the `@locks` collection is not yet known until we synchronize. It will also only work if object are strictly added to the collection. If we remove objects from @locks, that code could return a removed state in the `fetch` call. 

IMO this is an unnecessary optimization. This is nowhere near a hot path that could possibly from this optimization and it will not be thread safe if we eventually add objects removal from the `@locks` collection.

yaauie commented on logstash-core/lib/logstash/pipelines_registry.rb:
```diff
@@ -189,11 +258,5 @@ def select_pipelines(&optional_state_filter)
-
-    def get_lock(pipeline_id)
-      @locks.compute_if_absent(pipeline_id) do |k|
-        java.util.concurrent.locks.ReentrantLock.new
```
We are replacing a reentrant lock with a regular mutex. Is there any situation where a thread could attempt to acquire a lock that it already holds?

colinsurprenant replied:
Good catch. No spec has failed from this removal so it is unclear if this is in fact needed. But for the safety purposes I will assume there is/was a good reason for that and change the `Mutex` for a `Monitor` which **is** reentrant to make sure we preserve the same semantic here. I will also add a comment + followup issue to try to get to the bottom of that.

colinsurprenant replied:
After verification, I do not see any case where a reentrant lock is required. It would be useful only if any of the `create_pipeline`, `reload_pipeline` or `terminate_pipeline` methods could   call into each other which is not the case. IIRC at some point the actions were calling into themselves, for example the reload action was calling the stop and create actions, but this is not the case anymore.

colinsurprenant commented on logstash-core/lib/logstash/java_pipeline.rb:
```diff
@@ -176,21 +177,20 @@ def run
 
+      close
+    end
     @logger.debug("Pipeline has been shutdown", default_logging_keys)
```
No. No caller is using the `run` method exit code and no spec rely on that either. I believe this is an artefact of the Ruby Pipeline transposition and at some point it did have a meaning   but not anymore.

colinsurprenant commented on logstash-core/lib/logstash/java_pipeline.rb:
```diff
@@ -309,9 +316,17 @@ def resolve_cluster_uuids
+        # this is a worker thread termination
+        # delete it from @worker_threads so that wait_for_workers does not wait for it
+        @worker_threads.delete(terminated_thread)
+        raise("Worker thread terminated in pipeline.id: #{pipeline_id}")
```
So 2 things:
- Having a worker thread exception has always been considered a fatal at pipeline level. In the single pipeline scenario a worker thread exception has always been stopping logstash altogether. Before this PR a single worker exception in a multi pipeline scenario would ALSO crash logstash completely but that does not make sense and instead only the pipeline with the failed worker should be terminated. 
- When this is raised here, it is rescue'd in the `start` method. The calling chain is `start` -> `run` -> `monitor_inputs_and_workers`.  
https://github.com/elastic/logstash/blob/e99f37dc7b863ea23686ebfc4b8f926e969aaa23/logstash-core/lib/logstash/java_pipeline.rb#L123-L137

colinsurprenant replied:
I believe that raising here is correct in the sense that -it is- an exception to have a worker termination and that exception will be propagated and dealt with up the stack accordingly, otherwise a worker termination will simply terminate the pipeline in a non-exceptional way, just like when the inputs terminate which is not IMO correct.

colinsurprenant commented on logstash-core/lib/logstash/pipelines_registry.rb:
```diff
@@ -17,40 +17,99 @@
+    end
+
+    def pipeline
+      @lock.synchronize { @pipeline }
```
Yes, the pipeline object has been and is considered thread-safe itself. We only  want to guard against concurrent mutations in the state object itself.

colinsurprenant commented:
Thanks @yaauie for the review - I believe I clarified everything and pushed a fix for an oversight you pointed out.

colinsurprenant commented:
Jenkins test this

colinsurprenant commented:
jenkins test this

colinsurprenant commented on logstash-core/lib/logstash/java_pipeline.rb:
```diff
@@ -176,21 +177,20 @@ def run
 
-    shutdown_flusher
-    shutdown_workers
+      @logger.debug("Input plugins stopped! Will shutdown filter/output workers.", default_logging_keys)
```
Good catch.

colinsurprenant commented on logstash-core/lib/logstash/java_pipeline.rb:
```diff
@@ -125,14 +126,14 @@ def start
           "pipeline.sources" => pipeline_source_details)
         logger.error("Pipeline aborted due to error", pipeline_log_params)
       ensure
         @finished_execution.make_true
+        close
```
agree. will also trap exception in close because we must guarantee that `@finished_execution` is made true regardless of any exception.

colinsurprenant commented:
Thanks for the review @jsvd - I pushed some modifications and also answered some of your comments.

colinsurprenant commented:
@jsvd @yaauie anything else needed to move forward?

colinsurprenant commented:
@jsvd 

- I restructured the commits, you can now look at the diffs per commit for related changes. Let me know if you'd also prefer separate PRs.

> can you confirm that #1203 addresses the initial issue we saw with 3 pipelines, in that wasn't due to a slow pipeline initializing, but the special case where adding the monitoring source loader caused a convergence before the source was added, leading to an inconsistent state. 

You meant #12034 ^^ - Yes and no. That PR addressed the double convergence which was not required and removing that double convergence *practically* solved the problem **for that use-case**. The root problem is really that the `PipelineRegistry` allowed such an overlap by not accounting for a pipeline in its initialization phase. Without that *proper* fix this problem could resurface in other contexts. 

> Your other PR #12023 completes the fix where a crashed pipeline would keep the file lock.

That PR is not related to «the initial issue we saw with 3 pipelines» but addresses the case where any exception in `Queue.open()` would not release the lock and resulting in a "The queue failed to obtain exclusive access" exception when reloading the pipeline.  

For «the initial issue we saw with 3 pipelines» the "The queue failed to obtain exclusive access" exception was related to the fact that LS was trying to start a pipeline already in the process of being started.

jsvd commented:
I'm concerned about the user experience when an error happens to a worker thread.
As an exercise I introduced a "raise Exception" in the `def filter` of the logstash ruby filter then compared the errors we output in master compared to this PR:

master:

```
~/elastic/logstash master*
❯ bin/logstash -e "filter { ruby { code => 'raise Exception' } }" 
[2020-07-03T17:47:36,097][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"8.0.0", "jruby.version"=>"jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 OpenJDK 64-Bit Server VM 11.0.5+10 on 11.0.5+10 +indy +jit [darwin-x86_64]"}
[2020-07-03T17:47:37,278][INFO ][org.reflections.Reflections] Reflections took 37 ms to scan 1 urls, producing 22 keys and 45 values 
e[2020-07-03T17:47:37,695][INFO ][logstash.javapipeline    ][main] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>8, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>1000, "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x30822212 run>"}

[2020-07-03T17:47:38,409][INFO ][logstash.javapipeline    ][main] Pipeline Java execution initialization time {"seconds"=>0.7}
[2020-07-03T17:47:38,450][INFO ][logstash.javapipeline    ][main] Pipeline started {"pipeline.id"=>"main"}
The stdin plugin is now waiting for input:
[2020-07-03T17:47:38,517][INFO ][logstash.agent           ] Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
warning: thread "[main]>worker5" terminated with exception (report_on_exception is true):
java.lang.IllegalStateException: org.jruby.exceptions.Exception: (Exception) Exception
	at org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:101)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(jdk/internal/reflect/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)
	at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:441)
	at org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:305)
	at Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:282)
	at org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)
	at java.lang.Thread.run(java/lang/Thread.java:834)
Caused by: org.jruby.exceptions.Exception: (Exception) Exception
	at Users.joaoduarte.elastic.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_ruby_minus_3_dot_1_dot_5.lib.logstash.filters.ruby.filter(/Users/joaoduarte/elastic/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-ruby-3.1.5/lib/logstash/filters/ruby.rb:81)
	at Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.filters.base.do_filter(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:159)
	at Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:178)
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1809)
	at Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.filters.base.multi_filter(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/filters/base.rb:175)
	at org.logstash.config.ir.compiler.AbstractFilterDelegatorExt.multi_filter(org/logstash/config/ir/compiler/AbstractFilterDelegatorExt.java:134)
	at Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:282)
[2020-07-03T17:47:38,624][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>java.lang.IllegalStateException: org.jruby.exceptions.Exception: (Exception) Exception, :backtrace=>["org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:101)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke(jdk/internal/reflect/NativeMethodAccessorImpl.java:62)", "jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:441)", "org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:305)", "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:282)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)", "java.lang.Thread.run(java/lang/Thread.java:834)"]}
[2020-07-03T17:47:38,635][ERROR][org.logstash.Logstash    ] java.lang.IllegalStateException: Logstash stopped processing because of an error: (SystemExit) exit
```

And this PR:

```
~/elastic/logstash pr/12019* 8s
❯ bin/logstash -e "filter { ruby { code => 'raise Exception' } }"                             
[2020-07-03T17:42:06,210][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"8.0.0", "jruby.version"=>"jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 OpenJDK 64-Bit Server VM 11.0.5+10 on 11.0.5+10 +indy +jit [darwin-x86_64]"}
[2020-07-03T17:42:07,312][INFO ][org.reflections.Reflections] Reflections took 27 ms to scan 1 urls, producing 22 keys and 45 values 
[2020-07-03T17:42:07,693][INFO ][logstash.javapipeline    ][main] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>8, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>1000, "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x1cf010da run>"}
[2020-07-03T17:42:08,340][INFO ][logstash.javapipeline    ][main] Pipeline Java execution initialization time {"seconds"=>0.63}
[2020-07-03T17:42:08,402][INFO ][logstash.javapipeline    ][main] Pipeline started {"pipeline.id"=>"main"}
The stdin plugin is now waiting for input:
[2020-07-03T17:42:08,443][INFO ][logstash.agent           ] Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
[2020-07-03T17:42:08,628][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600}
e
[2020-07-03T17:42:40,491][ERROR][logstash.javapipeline    ][main] Pipeline worker error, the pipeline will be stopped {:pipeline_id=>"main", :error=>"org.jruby.exceptions.Exception: (Exception) Exception", :exception=>Java::JavaLang::IllegalStateException, :backtrace=>["org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:101)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke(jdk/internal/reflect/NativeMethodAccessorImpl.java:62)", "jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:441)", "org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:305)", "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:291)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)", "java.lang.Thread.run(java/lang/Thread.java:834)"], :thread=>"#<Thread:0x1cf010da sleep>"}
[2020-07-03T17:42:40,592][ERROR][logstash.javapipeline    ][main] Pipeline worker error {:pipeline_id=>"main", :exception=>#<RuntimeError: Worker thread terminated in pipeline.id: main>, :backtrace=>["/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:336:in `monitor_inputs_and_workers'", "/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:191:in `run'", "/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:134:in `block in start'"], "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x1cf010da run>"}
[2020-07-03T17:42:40,698][INFO ][logstash.runner          ] Logstash shut down.
```

If you notice, in master we show the exception's underlying cause, and reference the ruby.rb file:

```
Caused by: org.jruby.exceptions.Exception: (Exception) Exception
	at Users.joaoduarte.elastic.logstash.vendor.bundle.jruby.$2_dot_5_dot_0.gems.logstash_minus_filter_minus_ruby_minus_3_dot_1_dot_5.lib.logstash.filters.ruby.filter(/Users/joaoduarte/elastic/logstash/vendor/bundle/jruby/2.5.0/gems/logstash-filter-ruby-3.1.5/lib/logstash/filters/ruby.rb:81)
```

But that is gone in this PR.
Also, we'd now print two very similar error messages, and I don't see the benefit of having both:

```
[2020-07-03T17:42:40,491][ERROR][logstash.javapipeline    ][main] Pipeline worker error, the pipeline will be stopped {:pipeline_id=>"main", :error=>"org.jruby.exceptions.Exception: (Exception) Exception", :exception=>Java::JavaLang::IllegalStateException, :backtrace=>["org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:101)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)", "jdk.internal.reflect.NativeMethodAccessorImpl.invoke(jdk/internal/reflect/NativeMethodAccessorImpl.java:62)", "jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:566)", "org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:441)", "org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:305)", "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.java_pipeline.start_workers(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:291)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)", "java.lang.Thread.run(java/lang/Thread.java:834)"], :thread=>"#<Thread:0x1cf010da sleep>"}
[2020-07-03T17:42:40,592][ERROR][logstash.javapipeline    ][main] Pipeline worker error {:pipeline_id=>"main", :exception=>#<RuntimeError: Worker thread terminated in pipeline.id: main>, :backtrace=>["/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:336:in `monitor_inputs_and_workers'", "/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:191:in `run'", "/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/java_pipeline.rb:134:in `block in start'"], "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x1cf010da run>"}
```

colinsurprenant commented:
@jsvd Good catch on the logging. Note that the reason that the original exception was printed was because of the `report_on_exception` setting and not an explicit logging from the pipeline, that  exception detail would actually be lost when/if logging to file for example because it is a console only log and not going through log4j. 

I pushed 2 commits:

- I removed the redundant logging and I fixed the exception logging to provide the real exception cause and not the `IllegalStateException` wrapper.

- I added a spec to test on the correct termination of a pipeline upon a worker exception.

colinsurprenant commented:
@jsvd I added a final spec which tests that the pipeline registry creates a loading state before executing the create block so that a new pipeline in its initialization phase exists in the registry while it is in the process of initializing. This spec does not pass on master since the state was only **after** the create block was executed. 

I will leave these last commits as-is so you cal review them separately to assess them individually. Once approved I will squash.

Let me know if you have any more concerns.

jsvd commented on logstash-core/spec/logstash/pipelines_registry_spec.rb:
```diff
@@ -101,6 +101,39 @@
+        # stall here until subject.create_pipeline has been called and it entered the create block
+        wait_start_create_block.pop
+
+        expect(slow_initializing_pipeline).not_to receive(:finished_execution?)
```
This is rather complicated sequence to test a rather complicated scenario.
what do you think we can do to help readers understand this test quicker?

colinsurprenant replied:
added more comments

jsvd commented on logstash-core/spec/logstash/pipelines_registry_spec.rb:
```diff
@@ -101,6 +101,39 @@
+      let (:slow_initializing_pipeline) { double("slow_initializing_pipeline") }
+      let (:pipeline2) { double("pipeline2") }
+
+      it "should create a loading state before calling the create block" do
```
The test description can be confusing since it mentions that there should be a loading state, but then there's no other reference to asserting that the state is correct.

colinsurprenant replied:
added more comments. the only way to test without adding more API just for testing is to use indirect testing.

jsvd commented on logstash-core/spec/logstash/java_pipeline_spec.rb:
```diff
@@ -226,6 +237,38 @@ def flush(options)
+    end
+
+    it "does not raise in the main thread, terminates the run thread and finishes execution" do
+      subject.start
```
since the issue was that before the exception would bubble up,  maybe we should define the test in the negative. for example:

```ruby
it "does not raise in the main thread" do
  expect { subject.start && subject.shutdown }.to_not raise_error
end
```

colinsurprenant replied:
added to your suggestion.

colinsurprenant commented:
Rebased to 2 distinct commits; once green will merge. 
Thanks @yaauie @jsvd for the reviews.

colinsurprenant commented:
Also merged in 7.8 via #12109

colinsurprenant commented:
#12307 fixes a missing piece where the inputs are not stopped when terminating the pipeline upon a worker error.
